src/app/icd/server/ICDManager.cpp

```diff
@@ -234,15 +240,87 @@ bool ICDManager::CheckInMessagesWouldBeSent()
             }
 
             // At least one registration would require a Check-In message
-            VerifyOrReturnValue(mSubInfoProvider->SubjectHasActiveSubscription(entry.fabricIndex, entry.monitoredSubject), true);
+            VerifyOrReturnValue(!shouldCheckInMsgsBeSentFunction(entry.fabricIndex, entry.monitoredSubject), true);
         }
     }
 
     // None of the registrations would require a Check-In message
     return false;
 }
 
-void ICDManager::TriggerCheckInMessages()
+/**
+ * ShouldCheckInMsgsBeSentAtActiveModeFunction is used to determine if a Check-In message is required for a given registration.
+ * Due to how the ICD Check-In use-case interacts with the persistent subscription and subscription timeout resumption features,
+ * having a single implementation of the function renders the implementation very difficult to understand and maintain.
+ * Because of this, each valid feature combination has its own implementation of the function.
+ */
+#if CHIP_CONFIG_PERSIST_SUBSCRIPTIONS
+#if CHIP_CONFIG_SUBSCRIPTION_TIMEOUT_RESUMPTION
+/**
+ * @brief Implementation for when the persistant subscription and subscription timeout resumption feature are present.
+ *        Function checks that there no active or persisted subscriptions for a given fabricIndex or subjectID.
```

Comment:
It might be worth writing down _why_ this is the right thing to check in this configuration, so the next person to come along (us in 6 months) doesn't have to re-derive it.